usr/src/cmd/altexec/altexec.c

```diff
@@ -0,0 +1,180 @@
+/*
+ * This file and its contents are supplied under the terms of the
+ * Common Development and Distribution License ("CDDL"), version 1.0.
+ * You may only use this file in accordance with the terms of version
+ * 1.0 of the CDDL.
+ *
+ * A full copy of the text of the CDDL should have accompanied this
+ * source.  A copy of the CDDL is also available via the Internet at
+ * http://www.illumos.org/license/CDDL.
+ */
+
+/*
+ * Copyright 2023 MNX Cloud, Inc.
+ */
+
+/*
+ * Use a dedicated C program to exec() something in the pkgsrc paths.
+ */
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <assert.h>
+#include <errno.h>
+
+/* Remap table for when we have commands with than one name (e.g. tcsh/csh). */
+static struct {
+	const char *bin_name;
+	const char *bin_needs;
+} remap_table[] = {
+	{ "csh", "tcsh" },
+	/* login(1) invokes shells as -$SHELL. Cover that here. */
+	{ "-csh", "tcsh" },
+	{ "-tcsh", "tcsh" },
+	{ "-zsh", "zsh" },
+	{ NULL, NULL}
+};
+
+static const char *
+remap_bin(const char *base_bin)
+{
+	uint_t index;
+
+	for (index = 0; remap_table[index].bin_name != NULL; index++) {
+		/* Can use strcmp() because remap_table entries are bounded. */
+		if (strcmp(remap_table[index].bin_name, base_bin) == 0)
+			return (remap_table[index].bin_needs);
+
+	}
+	return (base_bin);
+}
+
+
+/*
+ * List the pkgsrc paths in desired search order. We are ignoring PATH
+ * intentionally.
+ */
+static const char *pkgsrc_paths[] = {
+	"/opt/local/bin",
+	"/opt/local/sbin",
+	"/opt/tools/bin",
+	"/opt/tools/sbin",
+	NULL
+};
+
+/*
+ * This returns an allocated string.  Since we're exec()-ing or exit()-ing we
+ * really don't need to worry about leaks, however.
+ */
+static char *
+generate_pkgsrc_path(const char *desired_bin)
+{
+	const char *path = NULL;
+	char *binpath;
+	uint_t index;
+
+	for (index = 0; pkgsrc_paths[index] != NULL; index++) {
+		int dirfd;
+		int cmdfd;
+
+		/*
+		 * NOTE: Any of dirfd or cmdfd that has an actual file
+		 * descriptor from open(2) MUST be closed before this loop
+		 * either exits or iterates.
+		 */
+
+		dirfd = open(pkgsrc_paths[index], O_RDONLY);
+		if (dirfd == -1) {
+			/* XXX KEBE ASKS, print diagnostics? */
```

Comment:
I was really unsure about diagnostics. I think that in general we should not print anything if there is something that we're going to exec later.

But, here's a few things I thought of...

* to syslog
* only when `-x` is passed (enables tracing in all shells). It would need to be supported anywhere in the args, and we need to make sure it's also passed to the `exec`'d program
* with static dtrace probes

Reply:
I make some judgement calls.  Two of them are standard-operating practices (searching paths).  The third is asprintf(3C) failing, which is worth spewage to stdout.